core/runtime/src/main/java/io/quarkus/runtime/StartupContext.java

```diff
@@ -21,21 +21,29 @@ public class StartupContext implements Closeable {
     // this is done to distinguish between the value having never been set and having been set as null
     private boolean lastValueSet = false;
     // the initial capacity was determined experimentally for a standard set of extensions
-    private final List<Runnable> shutdownTasks = new ArrayList<>(9);
-    private final List<Runnable> lastShutdownTasks = new ArrayList<>(7);
+    private final Deque<Runnable> shutdownTasks = new ConcurrentLinkedDeque<>();
+    private final Deque<Runnable> lastShutdownTasks = new ConcurrentLinkedDeque<>();
     private String[] commandLineArgs;
     private String currentBuildStepName;
 
     public StartupContext() {
         ShutdownContext shutdownContext = new ShutdownContext() {
             @Override
             public void addShutdownTask(Runnable runnable) {
-                shutdownTasks.add(runnable);
+                try {
+                    shutdownTasks.addFirst(runnable);
+                } catch (Throwable e) {
```

Comment:
What kind of exceptions do we need to handle here? I mean, it could throw `NullPointerException` but that's ok, or?

Reply:
But at this point you don't construct the task `Runnable` but merely add it to the collection...